allen-cell-animated/simularium-website
# Fix/ "Show/Hide All" checkbox bugs (Add tests for ModelPanel selectors)

Problem
=======
I wrote tests for the `ModelPanel` container's selectors and found some bugs having to do with the "Hide/Show All" checkbox. So I fixed the bugs too.

Repro for bugs:
1. Go to https://staging.simularium.allencell.org/viewer?trajFileName=endocytosis.simularium
2. Uncheck an agent in the left panel. Notice that the "Hide/show All" checkbox at the top of the list changes from a check mark to a hyphen (["indeterminate" checkbox state](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate)) (Correct behavior)
3. Now go to https://staging.simularium.allencell.org/viewer?trajFileName=pc4covid19.simularium
4. Uncheck "Dendritic cell" in the left panel. Notice that the "Hide/show All" checkbox at the top changes from a check mark to a blank checkbox (Incorrect behavior -- this checkbox should only be blank if nothing on the list is checked)
5. Now click on the arrow next to "Dendritic cell" to unfurl its states
6. Check a box on one of the dendritic cell's states (like "G0/G1 phase") to make it visible. Notice that the "Hide/show all" checkbox changed to a check mark (Incorrect behavior -- this checkbox should only be checked if everything on the list is checked)

Solution
========
1. I wrote tests for the `ModelPanel` selectors, including the selector that determines whether the "Show/Hide All" checkbox should be in an indeterminate state or not
2. The old selector `getCheckboxAllIsIntermediate` was failing I believe 3 of the new tests. I confirmed something was wrong by testing out the UI in staging (see repro steps above). 
3. I rewrote the above selector to fix the bugs and make the tests pass. It might not be the most optimal way to do it, but it seems to function correctly!

Also renamed some variables for clarity:
* `getCheckboxAllIsIntermediate` -> `getIsSharedCheckboxIndeterminate`
* `convertUITreeDataToSelectAll` -> `getSelectAllVisibilityMap`
* `convertUITreeDataToSelectNone` -> `getSelectNoneVisibilityMap`

And added an `npm run test-watch` script to package.json so we don't have to run `npm run test` after every change

## Type of change
Please delete options that are not relevant.

* Bug fix (non-breaking change which fixes an issue)
* This change requires updated or new tests

Steps to Verify:
----------------
1. `npm start` and check that the Hide/Show All checkbox behaves correctly now
2. `npm run test` - taken care of by GitHub Actions check

Keyfiles (delete if not relevant):
-----------------------
1. src/containers/ModelPanel/selectors.ts - Rewrite of `getCheckboxAllIsIntermediate`
2. src/containers/ModelPanel/selectors.test.ts - All new tests

Thanks for contributing!
